### ISSUE TITLE
Many warnings reading empty in-memory file

### ISSUE THREAD

**Issue by jimav** [Perl/perl5#19472]
If a read-write (open mode "+>") in-memory file is created and immediately read, many warnings appear. 

**Steps to reproduce**
```
#!/usr/bin/perl
use strict; use warnings; use v5.20;

open XX, "+>", "/tmp/testfile" or die $!;
seek XX,0,0 or die $!;
my @XXbuf = <XX>;  # no warnings (as expected)
say "Read (@XXbuf) from XX";

open YY, "+>", \(my $string) or die $!;
#print YY "stuff";
seek YY,0,0 or die $!;
my @YYbuf = <YY>;  # ******* MANY "Use of uninitialized value in <HANDLE>"
say "Read (@YYbuf) from YY";
```

**Expected behavior**
Expecting a silent EOF indicator, like a regular file.

P.S. truncate does not seem to work on in-memory files, but over-writing does.

**Perl configuration**
```
Summary of my perl6 (revision 5 version 30 subversion 0) configuration:
  
  Platform:
    osname=linux
    osvers=4.19.0
    archname=x86_64-linux-gnu-thread-multi
    uname='linux localhost 4.19.0 #1 smp debian 4.19.0 x86_64 gnulinux '
    config_args='-Dmksymlinks -Dusethreads -Duselargefiles -Dcc=x86_64-linux-gnu-gcc -Dcpp=x86_64-linux-gnu-cpp -Dld=x86_64-linux-gnu-gcc -Dccflags=-DDEBIAN -Wdate-time -D_FORTIFY_SOURCE=2 -g -O2 -fdebug-prefix-map=/build/perl-Wfb2Cd/perl-5.30.0=. -fstack-protector-strong -Wformat -Werror=format-security -Dldflags= -Wl,-Bsymbolic-functions -Wl,-z,relro -Dlddlflags=-shared -Wl,-Bsymbolic-functions -Wl,-z,relro -Dcccdlflags=-fPIC -Darchname=x86_64-linux-gnu -Dprefix=/usr -Dprivlib=/usr/share/perl/5.30 -Darchlib=/usr/lib/x86_64-linux-gnu/perl/5.30 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/x86_64-linux-gnu/perl5/5.30 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.30.0 -Dsitearch=/usr/local/lib/x86_64-linux-gnu/perl/5.30.0 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Duse64bitint -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Ud_ualarm -Uusesfio -Uusenm -Ui_libutil -Ui_xlocale -Uversiononly -DDEBUGGING=-g -Doptimize=-O2 -dEs -Duseshrplib -Dlibperl=libperl.so.5.30.0'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='x86_64-linux-gnu-gcc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    optimize='-O2 -g'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DDEBIAN -fwrapv -fno-strict-aliasing -pipe -I/usr/local/include'
    ccversion=''
    gccversion='9.3.0'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
"/tmp/j7" 164L, 10427C                                     1,19          Top
```

**Comment by Leont** [comment]
> If a read-write (open mode "+>") in-memory file is created and immediately read, many warnings appear.

Yeah, possibly PerlIO::scalar should convert an undefined value to an empty string.

> P.S. truncate does not seem to work on in-memory files, but over-writing does.

That is an entirely separate issue. I'm not sure it should, our abstraction kind of breaks there.

**Comment by Grinnz** [comment]
In memory files are required to be byte strings. You'll also get issues if you try to open it to a variable that contains wide characters, or is a reference. Maybe coercion is justified in this specific undef case.

**Comment by jimav** [comment]
There's nothing in the string.  Wide or octet.

**Comment by Grinnz** [comment]
Indeed, the problem is it is not a string. You get the same warnings for trying to print undef to a handle.

**Comment by Grinnz** [comment]
Somewhat poor example, sorry - basically what happened here is you told it to treat a scalar as a file, and that scalar's contents are undef, but it's impossible for a file's contents to be undef, so the abstraction doesn't sanely handle it.

**Comment by Grinnz** [comment]
Even more clearly: undef (the default state of a scalar value) and empty string are not the same, and you can avoid this by doing instead: `open my $fh, "+>", \(my $string = '')`

**Comment by jimav** [comment]
Okay, I understand.   As you say `open FH "+>", \(my $string="")` makes it work as intended, without warnings.

I am happy if this ticket is closed as user error.  Thank you for taking the time to explain.

**Comment by jkeenan** [comment]
> Okay, I understand. As you say `open FH "+>", \(my $string="")` makes it work as intended, without warnings.
> 
> I am happy if this ticket is closed as user error. Thank you for taking the time to explain.

Closed per OP consent.

**Comment by demerphq** [comment]
Hi James I am going to reopen this and assign it to myself. I think this is a terrible user experience and to me it violates autovivification expectations. I think if you use an undef var in this context it should autoviv to "". We can close it again if the community doesn't agree once I have a patch.

**Comment by demerphq** [comment]
Fixed in PR #19488

**Comment by jkeenan** [comment]
Given that we now have **4** BBC tickets opened due to https://github.com/Perl/perl5/commit/8b03aeb95ab72abdb2fa40f2d1196ce42f34708d, I am re-opening this issue so that the Perl Steering Committee can make a decision as to how we should proceed.

It should be noted that the original poster was satisfied with the explanation he was provided by several of our contributors and okayed the closing of the ticket. @demerphq re-opened the ticket and provided p.r. https://github.com/Perl/perl5/pull/19488.  While that p.r. was reviewed and okayed for merge, the fact that we have downstream breakage warrants reconsideration of the approach.  I don't have a personal opinion one way or the other on the problem or the fix, but wonder if perhaps this should go through a deprecation cycle.  The PSC should investigate.

@rjbs, @neilb, @leonerd 

Thank you very much.
Jim Keenan

**Comment by demerphq** [comment]
On Fri, 11 Mar 2022, 20:50 James E Keenan, ***@***.***> wrote:

> Given that we now have *4* BBC tickets opened due to 8b03aeb
> <https://github.com/Perl/perl5/commit/8b03aeb95ab72abdb2fa40f2d1196ce42f34708d>,
> I am re-opening this issue so that the Perl Steering Committee can make a
> decision as to how we should proceed.
>
> It should be noted that the original poster was satisfied with the
> explanation he was provided by several of our contributors and okayed the
> closing of the ticket. @demerphq <https://github.com/demerphq> re-opened
> the ticket and provided p.r. #19488
> <https://github.com/Perl/perl5/pull/19488>. While that p.r. was reviewed
> and okayed for merge, the fact that we have downstream breakage warrants
> reconsideration of the approach. I don't have a personal opinion one way or
> the other on the problem or the fix, but wonder if perhaps this should go
> through a deprecation cycle. The PSC should investigate.
>
> @rjbs <https://github.com/rjbs>, @neilb <https://github.com/neilb>,
> @leonerd <https://github.com/leonerd>
>

Let's just revert for now and pick it up in 5.37. I'm mindful of your
concern about the proximity to 5.36 release. It's not that important it
needs to go in now.

Yves

>

**Comment by demerphq** [comment]
On Sat, 12 Mar 2022 at 02:24, demerphq ***@***.***> wrote:
> On Fri, 11 Mar 2022, 20:50 James E Keenan, ***@***.***> wrote:
>>
>> Given that we now have 4 BBC tickets opened due to 8b03aeb, I am re-opening this issue so that the Perl Steering Committee can make a decision as to how we should proceed.
>>
>> It should be noted that the original poster was satisfied with the explanation he was provided by several of our contributors and okayed the closing of the ticket. @demerphq re-opened the ticket and provided p.r. #19488. While that p.r. was reviewed and okayed for merge, the fact that we have downstream breakage warrants reconsideration of the approach. I don't have a personal opinion one way or the other on the problem or the fix, but wonder if perhaps this should go through a deprecation cycle. The PSC should investigate.
>>
>> @rjbs, @neilb, @leonerd
>
>
> Let's just revert for now and pick it up in 5.37. I'm mindful of your concern about the proximity to 5.36 release. It's not that important it needs to go in now.

Pushed as https://github.com/Perl/perl5/pull/19527

I am good with working with the module authors to deal, or reverting.
Reverting is obviously simpler for the time being.

cheers,
yves
-- 
perl -Mre=debug -e "/just|another|perl|hacker/"
